### ISSUE TITLE
Kaocha uses technically illegal character (the period) in some keyword names

### ISSUE THREAD

**Issue by alysbrooks** [lambdaisland/kaocha#212]
@lread reported that @borkdude pointed out that `:kaocha.type/clojure.test` contains an illegal character in the name portion.

Grepping (well, `ag`ing) through the codebase to find more: 

```
ag "(:[a-zA-Z-.]+/[a-zA-Z-]+\.[a-zA-Z-]+|:?:[a-zA-Z-]+\.[a-zA-Z]+\s)" \
 --clojure  -o --nonumbers --nofilename | sort | uniq'

```

(This regex is imperfect but suffices for the initial bug report.)

These seem to fall into three categories:
* Ones used only in testing and documentation. These are easy to fix but are probably only worth fixing for completeness' sake.
    ```
   :bar.baz
    :foo.bar
    ```

* Ones that are both a Kaocha namespace and used as keywords in themselves. These should maybe be renamed (with some sort of conversion from the old names), but there's also a certain logic in, say, `:kaocha.result` containing keys `:kaocha/tests`.
    ```
    :kaocha.result
    :kaocha.testable
    ```

* Ones that refer to a type of test that's known by its namespace. Renaming these seems like it would mostly make things less intuitive on top of being extra work. We could avoid referring these as literals throughout the code base, but this has similar problems
    ```
    :kaocha.type/clojure.test
    :kaocha.type/spec.test
    ```

It's possible to create keywords with the `keyword` function (so `(keyword "bar.baz")`), and this doesn't violate the reader spec, which is limited to keyword literals. This might seem like an unimportant difference, but it does protect us against changes in the reader. I think keywords with these characters are always going to be possible to be created programatically because it gives more flexibility when deserializing external formats in general. In fact, EDN doesn't explicitly forbid periods in keyword names.

It's unclear that the reader will ever disallow these technically illegal characters. Clojure core developers briefly merged in a fix to the reader allowing leading numbers but [reverted it](https://clojure.atlassian.net/browse/CLJ-1252?focusedCommentId=16658) after it caused breakages. Presumably they will be reluctant to make similar fixes that could cause more issues.

In short, while these are not strictly correct, there are downsides to changing most of these and only a nebulous downside to keeping them. I was more leaning toward changing these when I started writing this issue than when I ended it, but it's still worth documenting my thought process, even if we decide not to fix this.

Remaining investigation:
- [x] Fix the regex to find keywords like `:kaocha.type/spec.test.check`
- [ ] Are there any plans to make the reader more strict?

**Comment by alysbrooks** [comment]
I'm leaning more against doing this change. But here's the improved regex:

`ag "(:[a-zA-Z-.]+\/[a-zA-Z-]+(\.[a-zA-Z-]+)+|:?:[a-zA-Z-]+(\.[a-zA-Z]+)+\s)" \
   --clojure  -o --nonumbers --nofilename | sort | uniq -c`

And here's the fuller list:

```     1 :bar.baz
      1 :foo.bar
      4 :kaocha.result 
      1 :kaocha.testable
     22 :kaocha.type/clojure.test
     11 :kaocha.type/spec.test.check
      4 :kaocha.type/spec.test.fdef
      5 :kaocha.type/spec.test.ns```

**Comment by plexus** [comment]
Yes, I think it's highly unlikely this will cause breakage, whereas us making these changes likely *will* cause breakage, so I'd say leave it and we'll cross that bridge if we ever get there.

**Comment by alysbrooks** [comment]
Apparently the no periods restriction was [relaxed on purpose](https://groups.google.com/g/clojure/c/CCuIp_bZ-ZM/m/9yP41gt7wuoJ), although it's possible they changed their minds on that. Maybe the core team will chime in on my Slack post but if not, I'm inclined to close this.

**Comment by alysbrooks** [comment]
Alex Miller advised someone not to use multiple slashes in keywords, which is a similar violation of what the reader is supposed to support: https://clojure.atlassian.net/browse/CLJ-1530?focusedCommentId=17285

Someone asked for periods in keywords to be officially allowed, but got no response from the core team: https://clojure.atlassian.net/browse/CLJ-889. As they point out, hiccup keywords contain spaces, so removing the ability to have periods in spaces would break a ton of Clojure code.

**Comment by borkdude** [comment]
More information on periods in keywords:

https://groups.google.com/g/clojure/c/CCuIp_bZ-ZM/m/9yP41gt7wuoJ

It seems we can safely use them. The docs need updating.

**Comment by plexus** [comment]
Wow, good thing someone found a post from 2008 to point out the docs are still wrong in 2021, also hopefully leading to a resolution to the JIRA ticket from 2011 last commented on in 2016 :D

**Comment by borkdude** [comment]
Never give up. https://github.com/clojure/clojure-site/issues/572
